system/metrics/sinks/prometheus/prometheusReporterService.cpp

```diff
@@ -0,0 +1,354 @@
+/*################################################################################
+#    HPCC SYSTEMS software Copyright (C) 2021 HPCC Systems®.
+#
+#    Licensed under the Apache License, Version 2.0 (the "License");
+#    you may not use this file except in compliance with the License.
+#    You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+################################################################################
+*/
+#include "prometheusReporterService.hpp"
+
+#include "jsocket.hpp"
+#include "jstream.ipp"
+
+#include "http/platform/httptransport.ipp"
+#include "espcontext.hpp"
+#include "httpsecurecontext.hpp"
+#include "htmlpage.hpp"
+
+using namespace hpccMetrics;
+
+PrometheusReporterService::PrometheusReporterService(const char * sinkname, const IPropertyTree * settings)
+{
+    m_metricsReporter = nullptr;
+    m_metricsSinkName = sinkname;
+    m_processing = false;
+    m_port = DEFAULT_PROMETHEUS_METRICS_SERVICE_PORT;
+    m_use_ssl = DEFAULT_PROMETHEUS_METRICS_SERVICE_SSL;
+    m_sslconfig = nullptr;
+
+    if (settings)
+    {
+        m_port = settings->getPropInt("@port", DEFAULT_PROMETHEUS_METRICS_SERVICE_PORT);
+        m_use_ssl = settings->getPropBool("@useSSL", DEFAULT_PROMETHEUS_METRICS_SERVICE_SSL);
+
+        settings->getProp("@serviceName", m_metricsServiceName);
+        if (m_metricsServiceName.isEmpty())
+            m_metricsServiceName = DEFAULT_PROMETHEUS_METRICS_SERVICE_NAME;
+
+        if(m_use_ssl)
+        {
+               m_sslconfig = settings->getBranch("sslConfig");
+#ifdef _USE_OPENSSL
+            if(m_sslconfig != nullptr)
+                m_ssctx.setown(createSecureSocketContextEx2(m_sslconfig, ServerSocket));
+            else
+                m_ssctx.setown(createSecureSocketContext(ServerSocket));
+#else
+           throw MakeStringException(-1, "PrometheusMetricsService: Cannot create SSL socket - OpenSSL not enabled in build");
+#endif
+        }
+    }
+
+    LOG(MCuserInfo, "PrometheusMetricsService created - port: '%i' uri: '/%s' ssl: '%s' sinkname: '%s'\n", m_port, m_metricsServiceName.str(), m_use_ssl ? "true" : "false", m_metricsSinkName);
+}
+
+int PrometheusReporterService::stop()
+{
+    m_processing = false; //this might need to be guarded
+    return 0;
+}
+
+bool PrometheusReporterService::isProcessing()
+{
+    return m_processing;
+}
+
+int PrometheusReporterService::start(MetricsReporter * pReporter)
```

Comment:
@ghalliday One of the reasons I hold off on giving the reporter to the sink until collection starts is that sink init takes place before the component has fully added all metrics. It was a way to enforce preventing the sink from accessing the reporter until all init had completed. If we don't want to enforce that, we can make the change, however, sinks should know to wait until collection is started before calling the reporter.

Reply:
@kenrowland We support dynamic metrics - so the sinks already need to support metrics being added after startCollection is called.  I can't think of any reason why they would call the reporter before startCollection() is called, and that can be a requirement if necessary.

Reply:
@ghalliday I agree that sinks must support metrics being added dynamically. I was looking for a way to enforce no sink calling the reporter until it had finished with its initialization. If a requirement on the sink is sufficient, then let me know and I will open a Jira to address it. I suggest we wait until after this PR is merged so that all sinks and the framework are handled in the same PR.